keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,220 @@
+            query[..., : self.rotary_dim],
+            query[..., self.rotary_dim:],
+        )
+        key_rot, key_pass = (
+            key[..., : self.rotary_dim],
+            key[..., self.rotary_dim:],
+        )
+
+        query_rot = self.rotary_embedding_layer(query_rot)
+        key_rot = self.rotary_embedding_layer(key_rot)
+
+        query = ops.concatenate((query_rot, query_pass), axis=-1)
+        key = ops.concatenate((key_rot, key_pass), axis=-1)
+
+        print("query shapes = ", ops.shape(query), query.shape)
+        print("key shapes = ", ops.shape(key), key.shape)
+
+        key = self._repeat_kv(key, self.num_key_value_groups)
+        value = self._repeat_kv(value, self.num_key_value_groups)
+
+        attention_output = self._compute_attention(query, key, value, attention_mask)
+
+        attention_output = ops.reshape(
+            attention_output,
+            [
+                ops.shape(attention_output)[0],
+                ops.shape(attention_output)[1],
+                self.hidden_dim,
+            ],
+        )
+
+        attention_output = self._output_dense(attention_output)
+
+        return attention_output
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+        if attention_mask is not None:
+            mask_expansion_axis = -3
+            for _ in range(
+                    len(attention_scores.shape) - len(attention_mask.shape)
+            ):
+                attention_mask = ops.expand_dims(
+                    attention_mask, axis=mask_expansion_axis
+                )
+        return self._softmax(attention_scores, attention_mask)
+
+    def _compute_attention(
+            self, query, key, value, attention_mask=None, training=None
+    ):
+        attention_scores = ops.einsum("aecd,abcd->acbe", key, query)
+
+        norm_factor = ops.sqrt(
+            ops.convert_to_tensor(self.attn_head_size, self.compute_dtype)
+        )
+
+        attention_scores /= norm_factor
+
+        attention_scores = self._masked_softmax(
+            attention_scores, attention_mask
+        )
+        attention_scores = self._attn_dropout_layer(
+            attention_scores, training=training
+        )
+        attention_output = ops.einsum(
+            "acbe,aecd->abcd", attention_scores, value
+        )
+
+        return attention_output
+
+    def _repeat_kv(self, hidden_states, num_repeats):
+        if num_repeats == 1:
+            return hidden_states
+
+        batch_size, seq_len, num_key_value_heads, head_dim = hidden_states.shape
```

Comment:
Hi @shivance maybe this should work
```suggestion
        hidden_states_shape = ops.shape(hidden_states)
        batch_size, seq_len = hidden_states_shape[0], hidden_states_shape[1]
        num_key_value_heads, head_dim = hidden_states_shape[2], hidden_states_shape[3]
```

Let me know if it solves the reshape problem or not.

Reply:
Damn. I totally forgot that we can't iterate over symbolic tensors. 
Thanks, this worked!